### PR TITLE
test: close the harness port race, and take four packages through condition and mutation coverage

### LINKED ISSUE #531: Unit test suite analysis before 2.8.0: per-test timing in cmd and internal, coverage to 100%, condition and mutation coverage, every finding fixed
Before 2.8.0 is tagged, a full analysis of the unit test suite, `cmd/` and `internal/` alike: where the time goes, test by test, and where coverage is short, measured with statement coverage, condition coverage and mutation testing. Every finding is fixed in the same pass, whether it was introduced yesterday or has been there since the first commit.

## Timing

- Measure the duration of every individual test and subtest (`go test -json` over `./cmd/... ./internal/...`, with `-count=1`), and produce a ranked report: slowest tests, slowest packages, and the share of the wall clock each takes. Keep the tool (a `cmd/` utility or a `make` target) so the report can be regenerated, and record the baseline in `docs/development/testing/`.
- For every test that consumes time it does not need: a sleep standing in for a condition, a real timeout where a short injected one would do, a fixture built per case instead of per package, a subprocess where an in-process seam would do, a network round-trip that a stub already covers. Rewrite it to assert the same thing faster, without weakening what it proves (the two previous rounds, https://github.com/jmrplens/gitlab-mcp-server/issues/484 and https://github.com/jmrplens/gitlab-mcp-server/issues/522, are the model: assert the mechanism, not the clock).
- The unit suite's wall clock in CI (the coverage job is the longest single command in the pipeline) goes down, and the report says by how much.

## Coverage

- Statement coverage: every package under `cmd/` and `internal/` to 100%, preexisting code included, using the seam pattern the packages already use (`cmd/server` is the exception recorded in https://github.com/jmrplens/gitlab-mcp-server/issues/359 only where a branch is genuinely unreachable; say which). The house rule is 100% on every touched package; this pass touches them all.
- Condition coverage with gobco (`make coverage-conditions PKG=…`) and mutation testing with gremlins (`make coverage-mutants PKG=…`) over the packages that carry logic (config, oauth, serverpool, subscriptions, toolutil, the tools orchestration, the audit commands): every lived mutant and every uncovered condition is either killed by a new test or recorded with the reason it cannot be.
- False passes: tests that cannot fail (no assertion, an assertion on the wrong value, a mock that answers whatever is asked) are found and fixed.

## Acceptance

- The timing report and the coverage figures are regenerated and committed in `docs/development/testing/`, with the before and after.
- `go test ./cmd/... ./internal/...` is faster than the baseline by a stated amount, statement coverage is 100% on every package except the recorded exceptions, and the mutation and condition reports carry no unexplained survivors.
- Every finding, current or preexisting, is fixed in this pass; nothing is filed for later.

### LINKED ISSUE #553: The HTTP end-to-end harness picks a port and releases it, so two parallel tests can collide on Windows
Seen on the Windows leg of a CI run:

```
--- FAIL: TestToolsList_DuringTheCatalogBuild_ReturnsTheWholeCatalog (0.27s)
    readiness_test.go:100: initialize answered 400, want 200:
    {"error":{"code":-32600,"message":"This deployment publishes no GitLab instance,
     so the GITLAB-URL header must name the one this request is for."}}
```

That test starts its server with `--gitlab-url` naming a fake GitLab, so the server it reached cannot be its own: a server started with an instance published never answers that message. The one that does is a server started with `--allow-any-gitlab-url`, which is what `withInstancePolicy` gives to almost every other test in the module. So the request went to another test's server.

The mechanism is `freePort` (`test/e2e/http/harness_test.go:127`): it listens on port zero, reads the port the kernel chose, closes the listener and returns the number. Between that close and the server binding, the port belongs to nobody, and every test in the module runs `t.Parallel()`. On Linux the window rarely loses; on Windows, where a just-closed port is handed out differently, two tests can be given the same number and the second server fails to bind or the first answers the second's requests.

The fake GitLab is not the problem: `httptest.NewServer` fills in its URL before the helper returns, so the flag was passed correctly.

Two ways out, either is fine:

- Hold the reservation: keep the listener open and hand the server the listener rather than the number. That needs the binary to accept a socket it did not open, which it does not today, so it is the larger change.
- Retry the bind: have `startServerOnPort` notice that the server failed to come up on the port it was given, pick another and try again, a small number of times. That closes the window without changing the binary, at the cost of a slower failure when a bind fails for a real reason.

Found while getting the race detector job green; the run before it and the run after both passed, so it is infrequent, which is exactly what makes it worth pinning rather than rerunning.

### PR DESCRIPTION
**Pull request by jmrplens**
Two issues, and what they turned into once I measured rather than guessed.

## A port the harness releases before it uses it (closes https://github.com/jmrplens/gitlab-mcp-server/issues/553)

The HTTP end-to-end harness asked the kernel for a free port, closed the listener and handed the number to the server it was about to start. Between those two moments the port belongs to nobody, so two tests running in parallel could be handed the same one, and on Windows the second server would bind it anyway and answer the first test's requests. That is worse than a failure: the test passes against somebody else's server.

The harness now retries on a taken port instead of trusting the number, and it watches the child process while it waits for `/health`, so a server that died during startup is reported as that rather than as a timeout.

## What the unit test suite actually costs, and what it does not check (closes https://github.com/jmrplens/gitlab-mcp-server/issues/531)

Statement coverage was already high enough to say nothing. So this measures two things it cannot see: condition coverage with `gobco`, which asks whether both halves of every branch were taken, and mutation testing with `gremlins`, which asks whether an assertion would notice if the code were wrong.

Four packages went through both and came out at 100% statements with their conditions closed: the GitLab client, configuration, OAuth and the server pool. The findings are the point, not the percentage:

- `CredentialRejected` in the GitLab client never returned `true` in any test. It exists to spot a 401 or 403 and drop a revoked pool entry, and a version reduced to `return false` passed the entire suite.
- The tool-surface selector accepts nineteen spellings and six had ever been evaluated. The capability-surface switch was reached only through its aliases, never through the canonical values an operator actually writes.
- `SatisfiesMinimum`'s whole reason for existing, that `api` supersedes `read_api`, was the untested half.
- Neither of the server pool's background sweeps had a test that ended it.
- Three bounds in configuration were asserted from one side only, including a size of zero that must be refused rather than accepted as a ceiling that then refuses every upload.

Forty-seven mutants survived and forty-three are equivalent, under three rules written down in the testing documentation: both sides agree at the boundary, a second guard makes the first unobservable, or the tool cannot instrument the construct. Where a branch was genuinely unreachable I simplified the code instead of writing a test that could not fail.

## Two tests that measured the machine

Windows CI caught both after the rebase, and both were green on Linux every time. A one-nanosecond credential ceiling is never exceeded where the monotonic clock advances in ticks, so `time.Since` answers zero and the case that exists to see a stale eviction saw none. The rate-limiter test was worse: its window was fifty milliseconds and it then filled four thousand keys, which on a slower runner outlasted the window, so the record the test is about was swept as lapsed before the test could lapse it and the next line dereferenced a map entry that was gone. Both now age their record deliberately instead of asking real time to pass.

## The columns I decided not to generate

`testing.md` is generated, and adding condition and mutation columns to it was the obvious next step. I measured what that would cost first: `gobco` is 4 to 283 seconds per package and `gremlins` 88 to 229, so four packages are about eleven minutes of mutation testing against six for the whole-tree coverage pass the generator already runs. `gobco` also cannot run at all on two of the ten target packages, because it copies every file ignoring build constraints and dies on the redeclaration. A number nobody can afford to regenerate goes stale while looking current, so the costs, the limitation and the three survivor-classification rules are written up by hand instead.

## Verification

Build and vet; the four packages driven to completion plus the server and dynamic packages; lint; both transport end-to-end modules vetted; the generators, with the testing reference and the stats refreshed from a full coverage run on this tree.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved reliability when OAuth tokens are refreshed or removed concurrently.
  - Prevented background maintenance tasks from stopping after a panic.
  - Improved handling of unusual GitLab responses and transport configurations.
  - Corrected server startup and health-check behavior, including dynamically assigned ports.

- **Documentation**
  - Added guidance on condition coverage and mutation testing.
  - Updated project metrics, test statistics, and coverage reports.

- **Tests**
  - Expanded coverage for configuration, authentication, server pooling, retries, and edge cases.

### CODE REVIEW

File: internal/serverpool/pool.go [hunk 1]
```diff
@@ -1144,6 +1172,35 @@ func tokenSuffix(token string) string {
...
+// credential's watchers. A panic in one of those is a bug in this process, but
+// losing the goroutine to it is worse than the bug is: the sweep stops
+// silently, and the pool then grows to its cap holding credentials nothing has
+// re-checked since startup. Deferred by all three goroutines, so they recover
+// the same way and the rule is stated once.
+func recoverSweep(ctx context.Context, sweep string) {
+	if r := recover(); r != nil {
+		slog.ErrorContext(ctx, "server pool: "+sweep+" goroutine panicked", "panic", r)
+	}
+}
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/serverpool/pool.go:1189]: Confirmed, and the recovery was contradicting its own comment: it says losing the goroutine is worse than the bug, and losing the goroutine is precisely what it did. The panic unwinds the whole goroutine before the recover sees it, so the ticker is gone and the pool then grows to its cap holding credentials nothing has re-checked since startup.

Each tick now runs inside `sweepTick`, which is the recovery around that tick alone; the goroutine keeps its own defer as the last resort, for a panic raised outside the tick. Both sweeps use it.

The test waited for one panic, which a recovery that ends the goroutine also delivers, so it proved only that the panic was caught. It now waits for two: the idle sweep panics on every in-use callback, and revalidation gets two refused entries so a second eviction can only come from a second tick. Both subtests fail against the previous code, which I checked by reverting `sweepTick` and running them. 816e7264.

File: test/e2e/http/harness_test.go [hunk 2]
```diff
@@ -251,27 +294,56 @@ func (w *lockedWriter) Write(p []byte) (int, error) {
...
 	t.Helper()
 	deadline := time.Now().Add(45 * time.Second)
 	for time.Now().Before(deadline) {
+		// Checked before the probe, because a server that has already given up
+		// leaves its port free for anything else to answer on, and a health
+		// check answered by somebody else is worse than one that fails: the
+		// test would go on to drive a server configured for another test.
+		select {
+		case <-exited:
+			return startupFailure(s.logs())
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply test/e2e/http/harness_test.go:374]: Right, and this one was a hang, not a slowdown: `awaitHealthy` takes the single value, the cleanup then blocks forever on `<-exited`, and the path that reaches it is exactly the one the retry existed for. The test would finish its assertions and the package would sit there until the timeout.

Fixed as you describe, by closing a `chan struct{}` after `cmd.Wait()` rather than sending: everything that asks whether the server is still there can ask, as many times as it likes. d8a5a115.

File: test/e2e/http/harness_test.go [hunk 3]
```diff
@@ -251,27 +294,56 @@ func (w *lockedWriter) Write(p []byte) (int, error) {
...
+}
+
+// startupFailure names why a server exited before it served, so that a port
+// lost to another test is retried and anything else is reported.
+func startupFailure(logs string) error {
+	if strings.Contains(logs, "address already in use") ||
+		strings.Contains(logs, "Only one usage of each socket address") {
+		return fmt.Errorf("%w. Output:\n%s", errPortTaken, logs)
+	}
+	return fmt.Errorf("the server exited before it served. Output:\n%s", logs)
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply test/e2e/http/harness_test.go:401]: Confirmed, and rather than tie readiness to the child I removed the guess the race is built on. `freePort` hands back a number instead of a listener, so the port belongs to nobody until the server binds it, and no amount of checking afterwards fixes a port that was never ours.

The harness now starts the server with `--http-addr=127.0.0.1:0` and reads the address back out of the server's own log. The kernel cannot hand the same port to two live listeners, and an address that came from this process's own output cannot name anybody else's, so the health check is bound to the managed child by construction and there is nothing left to retry.

That needed one change in the server: it logged nothing about a TCP listener, so with a port of 0 there was no way at all to learn which port a deployment got. It now logs the address it bound rather than the one it was asked for, which the unix socket branch has done since it existed.

One caller keeps a named port. Bare `--probe` discovers a listener by reading `--http-addr` off the other process's command line, and an ephemeral port is a `0` there, so discovery cannot find it. That is a real limitation of the flag rather than of the test, and it is now recorded where the flag is documented.

The whole HTTP end-to-end module passes on the new startup path. d8a5a115.

File: cmd/server/listen.go [hunk 4]
```diff
@@ -53,7 +53,16 @@ func isUnixSocketAddr(addr string) bool {
 // address is a path, a TCP port otherwise.
 func listenHTTP(ctx context.Context, addr string, socketMode os.FileMode) (net.Listener, error) {
 	if !isUnixSocketAddr(addr) {
-		return (&net.ListenConfig{}).Listen(ctx, "tcp", addr)
+		listener, err := (&net.ListenConfig{}).Listen(ctx, "tcp", addr)
+		if err != nil {
+			return nil, err
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply cmd/server/listen.go:58]: Valid, and it is the house rule: every error that leaves a function says what was being attempted. A bare `err` here reads as "address already in use" with no address in it, and the address is the whole content of that failure.

Carried into the next branch rather than this one. The stack under this PR is the 2.8.0 release and it is being merged now; the fix lands in the branch that follows it, together with the sweep-goroutine point below.

File: internal/serverpool/pool_test.go [hunk 5]
```diff
@@ -3103,3 +3105,672 @@ func TestEvictServer_DropsEveryEntryServedByThatServer(t *testing.T) {
...
+		pool.idleSweepInterval = time.Millisecond
+		if _, err := pool.GetOrCreateEntry("glpat-idle-panic", stubGitLabBase, nil); err != nil {
+			t.Fatalf("GetOrCreateEntry(): %v", err)
+		}
+
+		pool.StartIdleEviction(t.Context())
+
+		awaitSweepTicks(t, panicked,
+			"the idle sweep never ran",
+			"the idle sweep ran once and stopped: the panic took the ticker with it, so nothing is ever swept again")
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/serverpool/pool_test.go:3621]: Confirmed by reading it: `StartIdleEviction`'s goroutine can take a ready tick after `t.Context()` is cancelled, and `recoverSweep` then writes through the global logger that the next subtest has already replaced with an unsynchronized `strings.Builder`. That is a genuine race, and this repository now runs the detector on a schedule, so it would be found rather than tolerated.

Carried into the next branch, with the error-wrapping point above. The sweep needs its own context, cancelled and joined before the recovery subtest installs its logger.
